### DISCUSSION TITLE
Improving Character Rotation in VelocityDirection mode

### DISCUSSION THREAD

**Discussion by kriNon** [Sixze/ALS-Refactored#222] [General]
I've been trying to improve some issues related to Character Rotation in VelocityDirection mode. This work is closely related to issue #189.

There are some other minor issues with VelocityDirection mode, where it acts differently than expected. One major example of this occurs when you walk into a wall at a very slight angle. The wall will "push back" on the player, and the net movement of the character will be very small and tangential to the wall. This results in the resulting velocity being along the wall, where the player is trying to move towards the wall.

If we rotate the character in the direction of the character's velocity, we run into this issue.
What we really want is for the rotation to be based on the character's desired velocity - the velocity that the player would be moving at if no obstacles were in the way.

I wrote a quick and dirty implementation of the idea:

```
void UAlsCharacterMovementComponent::CalcVelocity(float DeltaTime, float Friction, bool bFluid,
	float BrakingDeceleration)
{
	const FVector StoreVelocity = Velocity;
	const FVector StoreAcceleration = Acceleration;
	
	Velocity = DesiredVelocity;
	
	
	// Apply acceleration
	if( !HasAnimRootMotion() && !CurrentRootMotion.HasOverrideVelocity() )
	{
		Super::CalcVelocity(DeltaTime, Friction, bFluid, BrakingDeceleration);
	}
	ApplyRootMotionToVelocity(DeltaTime);


	if( !HasAnimRootMotion() && !CurrentRootMotion.HasOverrideVelocity() && Velocity!=FVector::ZeroVector )
	{
		DesiredVelocity = Velocity;
	}
	
	//Restore Velocity and Acceleration
	Velocity = StoreVelocity;
	Acceleration = StoreAcceleration;
	
	Super::CalcVelocity(DeltaTime, Friction, bFluid, BrakingDeceleration);
}
```

I'd be really keen to hear if there is a better/cleaner (ideally simpler) way of implementing this.

**Comment by Sixze** [comment]
I added something not exactly what you are asking for, but something very similar. Check `bRotateTowardsDesiredVelocityInVelocityDirectionRotationMode` in the character settings.

Also, if you want to completely ignore the impact of penetration adjustments on velocity, you can simply remove the `Velocity = (UpdatedComponent->GetComponentLocation() - OldLocation) / timeTick;` line from `AlsCharacterMovementComponent.cpp`
